cryptographic_flows.md

```diff
@@ -0,0 +1,27 @@
+
+# Authenticate
+The local client inputs a user ID and a password and receives an "export key", which we will consider as a valid symmetric encryption key, as output. The key server receives as output some user-specific authentication material, which it can compare to its stored data to verify authentication.
+
+Both parties recieve an open channel as output, which satisfies mutual authentication, confidentiality, and integrity.
+
+
+# Generate a secret
+
+The local client component can generate a secret. Using a cryptographically secure pseudo-random number generator with a random seed, generate 256 bits of output.
+
+Generate a unique tag to associate with the secret (unique among secrets stored by the local client).
+
+# Generate and store a secret locally
+The local client component can run the [generate](generate-a-secret) protocol. Then, store the tag and the secret as a pair.
+
+# Generate and store a secret remotely
+The local client must
+1. Run the [authenticate](authenticate) protocol to retrieve an export key and an open channel.
+2. Run the [generate](generate-a-secret) protocol to get a tag and a secret.
+3. Use the export key to encrypt the secret.
+4. Send the tag and encrypted secret over the open channel.
+
+The key server generates a unique secret ID for the key by generating 256 bits of randomness and hashing it with the tag to get an ID for the secret (if the ID is not unique among the key server's stored IDs, try again).
+Then, store the user ID (from authentication), the secret ID and encrypted secret.
```

Comment:
I made #28 but the description feels too vague to me. I don't know what kinds of requirements this would entail (other than, like, "not modifiable by arbitrary parties"). Do you have more specific details about what kind of requirements should be included in that issue?

Reply:
I also am not clear on the vision for when to write TODOs in the code vs just having an open issue to fill in details are. In this case, I don't expect to change L25 when we address #28, so I'm inclined to not put a TODO here.

Reply:
My thought process is that whenever something is too vague for implementation, and you don't know how to write down the details, then more work (and a TODO with an issue) is required.

In this example, L25 is very vague on what "store" means, i.e., what requirements are there? Is it OK for the server to just put all this information in any old cleartext file? What choices need to be made in order to match our threat model and general design principles?

Basically, the specification will not be complete until we have a clear definition of what it means to "store" on the server side. So L25 will change, because it should reference these requirements.